code/game/objects/items/weapons/pneumaticCannon.dm

```diff
@@ -109,17 +113,53 @@
 		user.visible_message("<span class='danger'>[user] fires \the [src]!</span>", \
 				    		 "<span class='danger'>You fire \the [src]!</span>")
 	add_logs(user, target, "fired at", src)
+	var/turf/T = get_target(target, get_turf(src))
 	playsound(src.loc, 'sound/weapons/sonic_jackhammer.ogg', 50, 1)
-	for(var/obj/item/ITD in loadedItems) //Item To Discharge
-		loadedItems.Remove(ITD)
-		loadedWeightClass -= ITD.w_class
-		ITD.throw_speed = pressureSetting * 2
-		ITD.loc = get_turf(src)
-		ITD.throw_at(target, pressureSetting * 5, pressureSetting * 2,user)
+	fire_items(T, user)
 	if(pressureSetting >= 3 && user)
 		user.visible_message("<span class='warning'>[user] is thrown down by the force of the cannon!</span>", "<span class='userdanger'>[src] slams into your shoulder, knocking you down!")
 		user.Weaken(3)
 
+/obj/item/weapon/pneumatic_cannon/proc/fire_items(turf/target, mob/user)
+	switch(fire_mode)
+		if(PCANNON_FIREALL)
+			for(var/obj/item/ITD in loadedItems) //Item To Discharge
+				if(!throw_item(target, ITD, user))
+					break
+		if(PCANNON_FILO)
+			for(var/i = 0, i < throw_amount, i++)
```

Comment:
is it the two modes having nearly the exact same code and still being separated?